### PR TITLE
Vendor prometheus parser

### PR DESCRIPTION
**Pull request by fpetkovski**
This commit vendors in the promql parser from prometheus/prometheus. The purpose of this is to allow enable parsing new functions without interfering with the promql engine from Prometheus.

This code should ideally be temporary until upstream Prometheus adds support for xrate/xincrease and xdelta.

### CODE REVIEW

File: logicalplan/plan.go [hunk 1]
```diff
@@ -4,10 +4,12 @@
 package logicalplan
 
 import (
+	"fmt"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
 	"time"
```
Comment (saswatamcode) [inline logicalplan/plan.go:10]: This might need to be `goimports`ed

File: logicalplan/plan.go [hunk 2]
```diff
@@ -131,6 +133,112 @@ func traverseBottomUp(parent *parser.Expr, current *parser.Expr, transform func(
 	return true
 }
 
+// preprocessExpr wraps all possible step invariant parts of the given expression with
+// StepInvariantExpr. It also resolves the preprocessors.
+func preprocessExpr(expr parser.Expr, start, end time.Time) parser.Expr {
```
Comment (saswatamcode) [inline logicalplan/plan.go:138]: I'm guessing this had to be copied over, to work with `internal/prometheus/parser`? Might be worth adding a comment for that! 🙂

**Comment by fpetkovski** [comment]
All good points 👍 . Any idea how we can exclude the internal folder from faillint?

**Review by saswatamcode — APPROVED** [review]
LGTM!

Might need to fix the faillint command in Makefile to ignore the vendored parser.

File: logicalplan/plan.go (hunk 2 above)
Comment (fpetkovski) [reply logicalplan/plan.go:141]: Added

**Review by GiedriusS — APPROVED** [review]
:medal_military:

**Pull request merged**